File: StandardLibrary/runtimes/net/Extern/Time.cs
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Numerics;
using Microsoft.VisualBasic;
using Wrappers_Compile;
using ibyteseq = Dafny.ISequence<byte>;
using byteseq = Dafny.Sequence<byte>;
using icharseq = Dafny.ISequence<char>;
using charseq = Dafny.Sequence<char>;

namespace Time {

    public partial class __default {
        public static long CurrentRelativeTime()
        {
            var timespan = DateTime.Now - DateTime.MinValue;
            return (long) timespan.TotalSeconds;
        }

        public static _IResult<icharseq, icharseq> GetCurrentTimeStamp()
        {
            var utcTime = DateTime.UtcNow;
            try
            {
                // DateTime has format specifiers that make it easy to specify time stamp in different formats
                // using "o" as the format specifier gives us a time stamp in ISO8601
                // https://learn.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings#Roundtrip
                var timestamp = utcTime.ToString("yyyy'-'MM'-'dd'T'HH':'mm':'ss'.'ffffffK");
                return Result<icharseq, icharseq>.create_Success(charseq.FromString(timestamp));
            }
            catch (Exception e) when (e is ArgumentOutOfRangeException or FormatException)
            {
                return Result<icharseq, icharseq>.create_Failure(
                    charseq.FromString("Unable to generate timestamp in ISO8601 UTC format."));
            }
        }
    }
}

File: AwsEncryptionSDK/runtimes/rust/src/time.rs
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

#![deny(warnings, unconditional_panic)]
#![deny(nonstandard_style)]
#![deny(clippy::all)]

use crate::*;
use std::time::SystemTime;

impl crate::Time::_default {
    #[allow(non_snake_case)]
    pub fn CurrentRelativeTime() -> i64 {
        match SystemTime::now().duration_since(SystemTime::UNIX_EPOCH) {
            Ok(n) => n.as_secs() as i64,
            Err(_) => 0,
        }
    }

    #[allow(non_snake_case)]
    pub fn CurrentRelativeTimeMilli() -> i64 {
        match SystemTime::now().duration_since(SystemTime::UNIX_EPOCH) {
            Ok(n) => n.as_millis() as i64,
            Err(_) => 0,
        }
    }

    #[allow(non_snake_case)]
    pub fn GetCurrentTimeStamp() -> ::std::rc::Rc<
        _Wrappers_Compile::Result<
            ::dafny_runtime::Sequence<::dafny_runtime::DafnyCharUTF16>,
            ::dafny_runtime::Sequence<::dafny_runtime::DafnyCharUTF16>,
        >,
    > {
        let now_utc = chrono::Utc::now();
        let formatted = format!("{}", now_utc.format("%Y-%m-%dT%H:%M:%S%.fZ"));
        ::std::rc::Rc::new(
                _Wrappers_Compile::Result::Success{value :
                dafny_runtime::dafny_runtime_conversions::unicode_chars_false::string_to_dafny_string(&formatted)
                }
            )
    }
}
